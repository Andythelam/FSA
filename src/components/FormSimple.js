import React, { useEffect, useState } from 'react';
import { Link, Navigate } from 'react-router-dom';

const FormSimple = () => {
  const [form, setForm] = useState({
    expense23: '',
    expense22: '',
    expense21: '',
  });

  const handleChange = (e) => {
    setForm({
      ...form,
      [e.target.name]: e.target.value,
    });
  };

  const handleSubmit = () => {
    const data = {
      expense23,
      expense22,
      expense21,
    };
    fetch('http://localhost:3000/', {
      method: 'POST',
      body: JSON.stringify(data),
      headers: {
        'Content-Type': 'application/json',
      },
    })
      .then((res) => res.json())
      .then((expenses) => {
        console.log(expenses);
      })
      .catch((err) => {
        console.log(err);
      });
  };

  return (
    <div>
      <form className="bg-white p-8 rounded shadow-md w-full sm:w-96">
        <label className="text-lg font-semibold text-gray-700">
          What were your total average medical expenses in 2023? :
        </label>
        <input
          type="text"
          className="w-full border rounded p-2 mt-2 focus:outline-none focus:border-blue-500"
          name="expense23"
          onChange={handleChange}
        />
        <label className="text-lg font-semibold text-gray-700">
          What were your total average medical expenses in 2022? :
        </label>
        <input
          type="text"
          className="w-full border rounded p-2 mt-2 focus:outline-none focus:border-blue-500"
          name="expense22"
          onChange={handleChange}
        />
        <label className="text-lg font-semibold text-gray-700">
          What were your total average medical expenses in 2021? :
        </label>
        <input
          type="text"
          className="w-full border rounded p-2 mt-2 mb-3 focus:outline-none focus:border-blue-500"
          name="expense21"
          onChange={handleChange}
        />
        <button
          type="submit"
          className="flex w-full justify-center rounded-md bg-blue-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-blue-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-600"
          onClick={handleSubmit}
        >
          Submit
        </button>
      </form>
    </div>
  );
};

export default FormSimple;
