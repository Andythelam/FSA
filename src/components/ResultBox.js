import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';

const ResultBox = () => {
    
return (
    <div>ResultBox</div>
)
}


export default ResultBox;