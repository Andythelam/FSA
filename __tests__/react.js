import {render, screen, cleanup} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import React from 'react';

import FormSimple from '../src/components/FormSimple';
import ResultBox from '../src/components/ResultBox';

describe('Unit testing React components', () => {
  describe('Form simple', () => {
    let text;

    const mockFunction = jest.fn();

    beforeEach(() => {
      text = render(<FormSimple setServerResponse={mockFunction} />);
    });

    afterEach(() => {
      cleanup();
    });

    test('Renders three average medical expense forms', async () => {
      const forms = await screen.findAllByRole('textbox');
      expect(forms.length).toBe(3);
    });

    test('medical expense forms have correct label', async () => {
      const forms = await screen.findAllByRole('textbox');
      expect(forms[0].previousSibling).toHaveTextContent(
        'What were your average medical expenses in 2023?'
      );
      expect(forms[0].nextSibling).toHaveTextContent(
        'What were your average medical expenses in 2022?'
      );
      expect(forms[1].nextSibling).toHaveTextContent(
        'What were your average medical expenses in 2021?'
      );
    });

    test('Contains a submit button', async () => {
      const button = text.queryByRole('button');
      expect(button).toBeInTheDocument();
    });

    // bad test, buttons are just invoking a handlechange, not the passed down prop function
    // test('Functions are invoked on click', async () => {
    //   const button = await screen.findByRole('button');
    //   userEvent.click(button);
    //   expect(mockFunction).toHaveBeenCalledTimes(1);
    // });

    test('Snapshot test', () => {
      expect(text).toMatchSnapshot();
    });
  });

  describe('Result box', () => {
    let text;

    const props = [100, 200];

    afterEach(() => {
      cleanup();
    });

    // not sure why this test is failing lol
    // test('Renders no label if no figures provided', async () => {
    //   text = render(<ResultBox serverResponse={[]}/>);
    //   expect(text.getByText('Please enter numbers to calculate your estimate!')).toBe(false);
    // });

    test('Renders correct label if any NaN figures provided', async () => {
      text = render(<ResultBox serverResponse={[100, 'bye']} />);
      const label = await screen.getByText(
        'Please enter numbers to calculate your estimate!'
      );
      expect(label).toBeInTheDocument();
    });

    test('Renders correct label if valid figures provided', async () => {
      text = render(<ResultBox serverResponse={props} />);
      const label = await screen.getByText('Results:');
      expect(label).toBeInTheDocument();
      expect(label).toHaveStyle('font-weight: bold');
    });

    test('Properly renders amount 1', async () => {
      text = render(<ResultBox serverResponse={props} />);
      const label = await screen.getByTestId('amount1')
      const message = label.nextSibling
      const lossAmount = message.nextSibling.nextSibling
      expect(message).toHaveTextContent(
        'If you contributed the full amount, you likely lost'
      );
      expect(lossAmount).toHaveTextContent('$100');
      expect(lossAmount).toHaveStyle('font-weight: bold');
    });

    test('Properly renders amount 2', async () => {
      text = render(<ResultBox serverResponse={props} />);
      const label = await screen.getByTestId('amount2')
      const message = label.nextSibling
      const lossAmount = message.nextSibling.nextSibling
      expect(message).toHaveTextContent(
        'If you did not contribute, you likely lost'
      );
      expect(lossAmount).toHaveTextContent('$200');
      expect(lossAmount).toHaveStyle('font-weight: bold');
    });

    test('Snapshot test', () => {
      expect(text).toMatchSnapshot();
    })
  });
});