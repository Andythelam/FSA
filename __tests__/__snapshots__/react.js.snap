// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit testing React components Form simple Snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <form
          class="bg-white p-8 rounded shadow-md w-full sm:w-96"
        >
          <label
            class="text-lg font-semibold text-gray-700"
          >
            What were your average medical expenses in 2023?
          </label>
          <input
            class="w-full border rounded p-2 mt-2 focus:outline-none focus:border-blue-500"
            name="expense23"
            placeholder="$"
            type="text"
          />
          <label
            class="text-lg font-semibold text-gray-700"
          >
            What were your average medical expenses in 2022?
          </label>
          <input
            class="w-full border rounded p-2 mt-2 focus:outline-none focus:border-blue-500"
            name="expense22"
            placeholder="$"
            type="text"
          />
          <label
            class="text-lg font-semibold text-gray-700"
          >
            What were your average medical expenses in 2021?
          </label>
          <input
            class="w-full border rounded p-2 mt-2 mb-3 focus:outline-none focus:border-blue-500"
            name="expense21"
            placeholder="$"
            type="text"
          />
          <button
            class="flex w-full justify-center rounded-md bg-blue-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-blue-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-600"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <form
        class="bg-white p-8 rounded shadow-md w-full sm:w-96"
      >
        <label
          class="text-lg font-semibold text-gray-700"
        >
          What were your average medical expenses in 2023?
        </label>
        <input
          class="w-full border rounded p-2 mt-2 focus:outline-none focus:border-blue-500"
          name="expense23"
          placeholder="$"
          type="text"
        />
        <label
          class="text-lg font-semibold text-gray-700"
        >
          What were your average medical expenses in 2022?
        </label>
        <input
          class="w-full border rounded p-2 mt-2 focus:outline-none focus:border-blue-500"
          name="expense22"
          placeholder="$"
          type="text"
        />
        <label
          class="text-lg font-semibold text-gray-700"
        >
          What were your average medical expenses in 2021?
        </label>
        <input
          class="w-full border rounded p-2 mt-2 mb-3 focus:outline-none focus:border-blue-500"
          name="expense21"
          placeholder="$"
          type="text"
        />
        <button
          class="flex w-full justify-center rounded-md bg-blue-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-blue-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-blue-600"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Unit testing React components Result box Snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body />,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
